### PR TITLE
Allow users to use custom_objects to save custom Keras models

### PR DESCRIPTION
**Pull request by ankit-db**
## What changes are proposed in this pull request?
 
Add an optional custom_objects parameter to mlflow.keras.log_model and mlflow.keras.save_model that is semantically equivalent to the custom_objects parameter of keras.models.load_model() function. This dictionary is serialized and persisted with the MLflow model using CloudPickle. When the user loads an MLflow Keras model via mlflow.keras.load_model() or mlflow.pyfunc.load_model, we deserialize the custom_objects dictionary (if present) with CloudPickle and pass it to the keras.models.load_model() function.
 
## How is this patch tested?
 
New unit testing added with an example of constructing a custom model in Keras.
 
### Is this a user-facing change? 

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.
 
Add an optional custom_objects parameter to mlflow.keras.log_model and mlflow.keras.save_model that is semantically equivalent to the custom_objects parameter of keras.models.load_model() function, so custom Keras models can now be easily persisted with MLFlow.
 
### What component(s) does this PR affect?
 
- [ ] UI
- [ ] CLI 
- [x] API 
- [ ] REST-API 
- [ ] Examples 
- [ ] Docs
- [ ] Tracking
- [ ] Projects 
- [ ] Artifacts 
- [x] Models 
- [ ] Scoring 
- [ ] Serving
- [ ] R
- [ ] Java
- [x] Python

### How should the PR be classified in the release notes? Choose one:
 
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/keras.py [hunk 1]
```diff
@@ -89,7 +89,15 @@ def save_model(keras_model, path, conda_env=None, mlflow_model=Model()):
     os.makedirs(path)
     model_data_subpath = "model.h5"
     keras_model.save(os.path.join(path, model_data_subpath))
-
+    if custom_objects is not None:
```
Comment (dbczumar) [inline mlflow/keras.py]: If `custom_objects` is specified, we should also add `cloudpickle` as a dependency of the MLflow model. As a reference, the `mlflow.sklearn` module conditionally adds this dependency (see https://github.com/mlflow/mlflow/blob/26547aa98cc124ecdf8c5bd8cf80a6358da589b1/mlflow/sklearn.py#L129-L130).

We can achieve backwards compatibility when adding a parameter to the `mlflow.keras.get_default_conda_env()` method by using a keyword argument with a default value of `False` (reference: https://github.com/mlflow/mlflow/blob/26547aa98cc124ecdf8c5bd8cf80a6358da589b1/mlflow/sklearn.py#L39)

**Review by dbczumar — CHANGES_REQUESTED** [review]
@ankitmathur-db Nice progress! I left a few comments. The most significant one concerns the path resolution of the serialized model layers during the model loading process. Please let me know if you have any questions.

File: mlflow/keras.py [hunk 2]
```diff
@@ -144,24 +151,35 @@ def log_model(keras_model, artifact_path, conda_env=None, **kwargs):
...
+        with open(custom_objects_path, "rb") as in_f:
+            custom_objects = cloudpickle.load(in_f)
     from distutils.version import StrictVersion
-
     if StrictVersion(keras.__version__.split('-')[0]) >= StrictVersion("2.2.3"):
         # NOTE: Keras 2.2.3 does not work with unicode paths in python2. Pass in h5py.File instead
         # of string to avoid issues.
         with h5py.File(os.path.abspath(model_file), "r") as model_file:
-            return keras.models.load_model(model_file, **kwargs)
+            return keras.models.load_model(model_file, custom_objects=custom_objects, **kwargs)
```
[the remark this answers is not in the document]
Comment (ankitmathur-db) [reply mlflow/keras.py]: Good point - I'd missed that! Updated wrt to this.

File: mlflow/keras.py [hunk 3]
```diff
@@ -144,24 +158,42 @@ def log_model(keras_model, artifact_path, conda_env=None, **kwargs):
...
+    custom_objects_path = None
+    if os.path.isdir(model_path):
+        if os.path.isfile(os.path.join(model_path, _CUSTOM_OBJECTS_SAVE_PATH)):
+            custom_objects_path = os.path.join(model_path, _CUSTOM_OBJECTS_SAVE_PATH)
+        model_path = os.path.join(model_path, _MODEL_SAVE_PATH)
+    if custom_objects_path is not None:
+        import cloudpickle
+        with open(custom_objects_path, "rb") as in_f:
+            pickled_custom_objects = cloudpickle.load(in_f)
+            custom_objects.update(pickled_custom_objects)
```
Comment (dbczumar) [inline mlflow/keras.py]: This update ordering here will ignore user-specified `custom_objects` if they are already present in `pickled_custom_objects`. I think we should preferentially select the user-specified `"custom_objects"` dictionary instead by reversing the update order (e.g., `pickled_custom_objects.update(custom_objects)`).

Comment (dbczumar) [reply mlflow/keras.py]: Can we also add a test to ensure that user-specified custom objects are preferentially selected?

File: tests/keras/test_keras_model_export.py [hunk 4]
```diff
@@ -228,7 +306,10 @@ def test_model_load_succeeds_with_missing_data_key_when_data_exists_at_default_p
     can be loaded successfully. These models are missing the `data` flavor configuration key.
     """
     mlflow.keras.save_model(keras_model=model, path=model_path)
-
+    print(model_path)
```
Comment (dbczumar) [inline tests/keras/test_keras_model_export.py]: Nit: remove print statement

File: tests/keras/test_keras_model_export.py [hunk 5]
```diff
@@ -228,7 +306,10 @@ def test_model_load_succeeds_with_missing_data_key_when_data_exists_at_default_p
...
+    shutil.copyfile(
```
Comment (dbczumar) [inline tests/keras/test_keras_model_export.py]: Should this be a move instead of a copy?

File: mlflow/keras.py [hunk 6]
```diff
@@ -23,16 +23,21 @@
 from mlflow.utils.model_utils import _get_flavor_configuration
 
 FLAVOR_NAME = "keras"
+_CUSTOM_OBJECTS_SAVE_PATH = "custom_objects.cloudpickle"
```
Comment (dbczumar) [inline mlflow/keras.py]: Nit: Can we add inline comments documenting the purpose of these module-level strings variables?

File: mlflow/keras.py [hunk 7]
```diff
@@ -144,24 +158,42 @@ def log_model(keras_model, artifact_path, conda_env=None, **kwargs):
     >>>   mlflow.keras.log_model(keras_model, "models")
     """
     Model.log(artifact_path=artifact_path, flavor=mlflow.keras,
-              keras_model=keras_model, conda_env=conda_env, **kwargs)
+              keras_model=keras_model, conda_env=conda_env, custom_objects=custom_objects, **kwargs)
+
 
+def _save_custom_objects(path, custom_objects):
```
Comment (dbczumar) [inline mlflow/keras.py]: Can we document the semantics of `path`? Does `path` refer to the MLflow model path or the data directory path or the destination path for the pickled custom objects?

Perhaps it would be clearer if this method accepted the destination path for the custom objects directly; the caller can then construct the path and call `_save_custom_objects(os.path.join(path, custom_objects_subpath), ...)`.

File: mlflow/keras.py [hunk 8]
```diff
@@ -144,24 +158,42 @@ def log_model(keras_model, artifact_path, conda_env=None, **kwargs):
...
+    import cloudpickle
+    custom_objects_subpath = _CUSTOM_OBJECTS_SAVE_PATH
```
Comment (dbczumar) [inline mlflow/keras.py]: Nit: We can just use the module-level variable directly - no need to assign it to a function-local variable.

**Review by dbczumar — CHANGES_REQUESTED** [review]
@ankitmathur-db This is really close! Left a few more small comments (mostly nits!)

File: mlflow/keras.py (hunk 7 above)
Comment (ankitmathur-db) [reply mlflow/keras.py]: Took a pass at it without that - let me know if you still think it's necessary with the documentation

File: mlflow/keras.py (hunk 3 above)
Comment (ankitmathur-db) [reply mlflow/keras.py]: Great idea! Done

File: mlflow/keras.py [hunk 9]
```diff
@@ -118,8 +129,7 @@ def log_model(keras_model, artifact_path, conda_env=None, **kwargs):
                       contained in :func:`get_default_conda_env()`. If ``None``, the default
                       :func:`mlflow.keras.get_default_conda_env()` environment is added to
                       the model. The following is an *example* dictionary representation of a
-                      Conda environment::
-
+                      Conda environment:
```
Comment (dbczumar) [inline mlflow/keras.py]: Can we revert this change? The double colon signifies the beginning of a literal block in RST (http://docutils.sourceforge.net/docs/ref/rst/restructuredtext.html#literal-blocks).

File: mlflow/keras.py [hunk 10]
```diff
@@ -144,24 +154,47 @@ def log_model(keras_model, artifact_path, conda_env=None, **kwargs):
...
-              keras_model=keras_model, conda_env=conda_env, **kwargs)
+              keras_model=keras_model, conda_env=conda_env, custom_objects=custom_objects, **kwargs)
+
 
+def _save_custom_objects(path, custom_objects):
+    """
+    Save custom objects dictionary to a cloudpickle file so a model can be easily loaded later.
 
-def _load_model(model_file, **kwargs):
+    :param path: An absolute path that that points to the data directory within /path/to/model.
```
Comment (dbczumar) [inline mlflow/keras.py]: Nit: double "that" (`absolute path that that points`)

File: tests/keras/test_keras_model_export.py [hunk 11]
```diff
@@ -98,6 +146,47 @@ def test_model_save_load(model, model_path, data, predicted):
...
+
+
+def test_custom_model_save_respects_user_custom_objects(custom_model, custom_layer, model_path):
+    class DifferentCustomLayer():
+        def __init__(self):
+            pass
+    incorrect_custom_objects = {'MyDense': DifferentCustomLayer()}
+    correct_custom_objects = {'MyDense': custom_layer}
+    mlflow.keras.save_model(custom_model, model_path, custom_objects=incorrect_custom_objects)
+    model_loaded = mlflow.keras.load_model(model_path, custom_objects=correct_custom_objects)
```
Comment (dbczumar) [inline tests/keras/test_keras_model_export.py]: Can we also verify that calling `load_model` with `incorrect_custom_objects` throws an exception?

**Review by dbczumar — APPROVED** [review]
LGTM after 3 small comments are addressed! Thanks @ankitmathur-db!
